### ISSUE TITLE
RowDetailView way to tell when  reloaded

### ISSUE THREAD

**Issue by SatanEnglish** [6pac/SlickGrid#252]
@ghiscoding @6pac

Need a better way of handling the reload.

Currently inside function detailSelectionFormatter(row, cell, value, columnDef, dataContext)
I am calling a notify with a timeout to allow for the formater to finish doing it's job like so.

            setTimeout(function () {
                // trigger an event after toggling
                _self.onAfterDetailViewFormatterLoad.notify({
                    "grid": _grid,
                    "item": dataContext
                }, null, _self);
            }, 50);
Can you guys think of a better way of doing this?

Basically this is called every time you scroll the expanding view back into view.
Purpose of this is so that I can fire off scripts to re-setup js that no longer works once the view leaves the cache. (E.g they are no longer attached as the item left the DOM)

**Comment by ghiscoding** [comment]
Maybe you should try [Post Render](http://6pac.github.io/SlickGrid/examples/example10-async-post-render.html)

**Comment by SatanEnglish** [comment]
I've got a bunch of updates for this component that I'll have to push.
Will look into the Post Render not sure it will do what but I'll have a look.

Really should look at moving the details for the expanded off of the item or into a sub object to.
e.g. currently row._padding => row.expanderObj = { padding: 5, collapsed: false  }

**Comment by 6pac** [comment]
Sorry to not be of much help, but I've stayed out of this because I was a bit worried it might open up a can of worms. The issues are significantly complex. Inline editing is always tricky because it combines UI logic with scrolling. 
The Post Render and Post Render cleanup are designed for exactly this kind of situation (instantiation and cleanup of 'heavy' controls in the grid), but may not suit in their current form. Perhaps it's better to make them a bit more flexible rather than invent something else.

From a repo point of view I'm happy to pull whatever goes on within the plugin, and to accept minor core changes.  
To be blunt, I'm happy to support the work you do, but I don't really have time to jump in and be part of it! I can spend half an hour here and there, but not hours.

**Comment by ghiscoding** [comment]
@SatanEnglish 
I'm nearly done with my refactoring and testing of the updated code for the Row Detail View plugin. I added 2 new subscribe methods (`onRowOutOfVisibleRange` and `onRowBackToVisibleRange`) which as the name suggest notify the user when a row detail is out of visibility (from user's perspective, so if user can't see it anymore, then the `onRowOutOfVisibleRange` triggers and when the row comes back to user eye then `onRowBackToVisibleRange` triggers. 

So my question is, does that address/fix this open issue? If not, please provide more detail

**Comment by SatanEnglish** [comment]
@ghiscoding 
Assuming that the onRowBackToVisibleRange is fired off after the view is reloaded in then yes that sounds like it would do what is needed.

**Comment by ghiscoding** [comment]
No it's fired off **before** and we use it to trigger a reload of the view. For example, in other framework like Angular/Aurelia, we need to use the framework's own rendering engine (which is different in each framework). So I don't even need or use the `SaveViewDetail` that you created (I made it optional via a flag). 

I don't understand why you need to know when the view finished rendering. The template is a regular html template, and I assume you would load the view in a synchronous way without waiting for anything (or in the case of using other framework, then use the framework rendering engine to know when it's finished/ready).

**Comment by SatanEnglish** [comment]
I think I needed the view to be present before I ran the javascript that is attached to the DetailView. (It needs to be present before I call resizeCanvas on an internal grid for example)

**Comment by ghiscoding** [comment]
I think that will work with the new subscribe.

```javascript
rowDetail.onRowBackToVisibleRange,subscribe(function(e, args) {
  // do your re-setup
}
```

This is the signature of this new subscribe
```js
/**
 *    onRowBackToVisibleRange: Fired after the row detail gets toggled
 *      Event args:
 *        grid:         Reference to the grid.
 *        item:         Item data context
 *        rowIndex:     Index of the Row in the Grid
 *        expandedRows: Array of the Expanded Rows
 *        outOfVisibleRangeRows: Array of the Out of Visible Range Rows
 */
```

**Comment by SatanEnglish** [comment]
@ghiscoding
Found what I'm using it for.
When the detailView is opened make sure it fits on the screen
This is not necessarily from scrolling there is one other case where I needed it to fire.
This is when the view gets re-expanded for a second time it goes through and calls resize on all inner grids else you get warped column sizes all stuck to the left.

**Comment by ghiscoding** [comment]
No those 2 new subscribers are only meant to be used with expanded rows only. It will tell when they become visible or not, that is what they are for and only that. Why would you care about collapsed row? There's nothing to do with collapsed rows, at least on my side I'll never do anything with them

**Comment by SatanEnglish** [comment]
@ghiscoding 
As I said above this is I => S1 expanded => S2 collapse => S3 re-expand row.
Does the onRowBackToVisibleRange  fire off on S3??

**Comment by ghiscoding** [comment]
@SatanEnglish 
No it won't fire off, that's not it's job, what will fire is `onAfterRowDetailToggle`. 

The reason of these 2 new subscribers is for visibility purposes **only**, what's visible to the user in the viewport and when it becomes outside. You would use `onRowOutOfVisibleRange` when it's just about to be no longer visible (so for instance you might want to save something before it gets lost), and then re-render the view when it comes back visible with `onRowBackToVisibleRange`. This is what it's used for and nothing else. 

This is the code that you had written
```javascript
// Find the bottom most item that is likely to go off screen
var bottomMostItem = _dataView.getItemByIdx(end - 1);

// If we are a detailView and we are about to go out of cache view
if (bottomMostItem && bottomMostItem._parent !== undefined) {
   saveDetailView(bottomMostItem._parent);
}
```
I thought `onRowOutOfVisibleRange` would help in this situation... isn't it what you wanted to do? My new code does call the `saveDetailView()` before being out of visibility, because of your previous code (that I've put in this post). Did I misunderstand your code logic?

**Comment by SatanEnglish** [comment]
@SatanEnglish

> No it won't fire off, that's not it's job, what will fire is onAfterRowDetailToggle.
This answered my question here. All good i'll just have to put an extra subscribe in for this case.

The scrolling stuff seems to be fine!
